src/main/java/co/cask/hydrator/salesforce/SObjectDescriptor.java

```diff
@@ -0,0 +1,186 @@
+ * Can be obtained from SOQL query or from SObject name.
+ */
+public class SObjectDescriptor {
+
+  private final String name;
+  private final List<FieldDescriptor> fields;
+
+  /**
+   * Connects to Salesforce, gets describe result for the given sObject name and stores
+   * information about its fields into {@link SObjectDescriptor} class.
+   *
+   * @param name sObject name
+   * @param credentials Salesforce connection credentials
+   * @return sObject descriptor
+   * @throws ConnectionException in case of errors when establishing connection to Salesforce
+   */
+  public static SObjectDescriptor fromName(String name, AuthenticatorCredentials credentials)
+    throws ConnectionException {
+    PartnerConnection partnerConnection = SalesforceConnectionUtil.getPartnerConnection(credentials);
+    SObjectsDescribeResult describeResult = new SObjectsDescribeResult(
+      partnerConnection, Collections.singletonList(name));
+    List<FieldDescriptor> fields = describeResult.getFields().stream()
+      .map(Field::getName)
+      .map(FieldDescriptor::new)
+      .collect(Collectors.toList());
+    return new SObjectDescriptor(name, fields);
+  }
+
+  /**
+   * Parses given SOQL query and stores information about sObject name and its fields into
+   * {@link SObjectDescriptor} class.
+   *
+   * @param query SOQL query
+   * @return sObject descriptor
+   */
+  public static SObjectDescriptor fromQuery(String query) {
+    return SalesforceQueryParser.getObjectDescriptorFromQuery(query);
+  }
+
+  public SObjectDescriptor(String name, List<FieldDescriptor> fields) {
+    this.name = name;
+    this.fields = new ArrayList<>(fields);
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  /**
+   * Collects sObject names needed to be described in order to obtains field type information.
+   *
+   * @return list of sObject names
+   */
+  public Set<String> getAllParentObjects() {
+    Set<String> parents = fields.stream()
+      .filter(FieldDescriptor::hasParents)
+      .map(FieldDescriptor::getLastParent)
+      .collect(Collectors.toSet());
+
+    // add top level sObject for fields that don't have parents
+    parents.add(name);
+
+    return parents;
+  }
+
+  /**
+   * Collects all field names, for fields with parents includes parents separated by dot.
+   *
+   * @return list of field names
+   */
+  public List<String> getFieldsNames() {
+    return fields.stream()
+      .map(FieldDescriptor::getFullName)
+      .collect(Collectors.toList());
+  }
+
+  public List<FieldDescriptor> getFields() {
+    return fields;
+  }
+
+  @Override
+  public String toString() {
+    return "SObjectDescriptor{" + "name='" + name + '\'' + ", fields=" + fields + '}';
+  }
+
+  /**
+   * Contains information about field, including list of parents if present.
+   */
+  public static class FieldDescriptor {
+
+    private final String name;
+    private final List<String> parents;
```

Comment:
General question, how many layers of nesting is possible? Eg, it's possible to have o.Name. Is it possible to have o.Name.Y, or o.y.Name?

Reply:
In each specified relationship, no more than five levels can be specified in a child-to-parent relationship. For example, Contact.Account.Owner.FirstName (three levels).

https://developer.salesforce.com/docs/atlas.en-us.salesforce_app_limits_cheatsheet.meta/salesforce_app_limits_cheatsheet/salesforce_app_limits_platform_soslsoql.htm

Reply:
I see - let's add validation that only 5 levels can be specified then